habitat/core/dataset.py

```diff
@@ -236,20 +246,116 @@ def get_splits(
-            whole episodes
+            episodes: list of episodes.
+            cycle: if true, cycle back to first episodes when StopIteration.
+            shuffle_scene_when_cycle: if true, shuffle scene groups when cycle.
+            group_by_scene: if true, group episodes from same scene.
+            max_scene_repeat: threshold of how many episodes from the same
+                scene can be loaded consecutively. -1 for no limit
+            num_episode_sample: number of episodes to be sampled.
+                -1 for no sampling.
         """
-        if num_episodes == -1:
-            return
-        if num_episodes < -1:
-            raise ValueError(
-                f"Invalid number for episodes to sample: {num_episodes}"
+        # sample episodes
+        if num_episode_sample != -1:
+            if num_episode_sample < -1:
+                raise ValueError(
+                    f"Invalid episode number to sample: {num_episode_sample}"
+                )
+            episodes = np.random.choice(
+                episodes, num_episode_sample, replace=False
             )
-        self.episodes = np.random.choice(
-            self.episodes, num_episodes, replace=False
-        )
+        self.episodes = episodes
+        self.cycle = cycle
+        self.group_by_scene = group_by_scene
+        if group_by_scene:
+            num_scene_groups = len(
+                list(groupby(episodes, key=lambda x: x.scene_id))
+            )
+            num_unique_scenes = len(set([e.scene_id for e in episodes]))
+            if num_scene_groups >= num_unique_scenes:
+                self.episodes = sorted(self.episodes, key=lambda x: x.scene_id)
+        self.max_scene_repetition = max_scene_repeat
+        self.shuffle_scene_when_cycle = shuffle_scene_when_cycle
+        self.rep_count = 0
+        self._iterator = None
+        self.prev_scene_id = None
+        self._iterator = iter(self.episodes)
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        r"""
+        The main logic for handling how episodes will be iterated.
+        Returns:
+            next episode.
+        """
+        try:
+            next_episode = next(self._iterator)
+        except StopIteration:
+            if not self.cycle:
+                raise StopIteration
+            self._iterator = iter(self.episodes)
+            if self.shuffle_scene_when_cycle:
+                self._shuffle_iterator_scene()
+            next_episode = next(self._iterator)
+
+        if self.prev_scene_id == next_episode.scene_id:
+            self.rep_count += 1
+        if (
+            self.max_scene_repetition > 0
+            and self.rep_count >= self.max_scene_repetition - 1
+        ):
+            self._shuffle_iterator_scene()
+            self.rep_count = 0
+
+        self.prev_scene_id = next_episode.scene_id
+        return next_episode
+
+    def _shuffle_iterator_scene(self) -> None:
+        r"""
```

Comment:
Can you follow the same convention for docstrings as elsewhere, I think the comment starts from the first line.